ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/recovery/BasicActionExpiryScanner.java

```diff
@@ -0,0 +1,111 @@
+import com.arjuna.ats.internal.arjuna.common.UidHelper;
+
+public abstract class BasicActionExpiryScanner<T extends BasicAction> implements ExpiryScanner {
+
+    private static final Duration EXPIRY_TIME;
+    private static final DateTimeFormatter TIME_FORMAT = DateTimeFormatter.ofPattern("EEE, d MMM yyyy HH:mm:ss");
+
+    private final RecoveryStore recoveryStore;
+    private final Class<T> typeClass;
+    private final String typeName;
+    private final String movedTypeName;
+
+    static {
+        EXPIRY_TIME = Duration.ofHours(recoveryPropertyManager.getRecoveryEnvironmentBean().getExpiryScanInterval());
+    }
+
+    public BasicActionExpiryScanner(T typeInstance, String movedSubFolder) {
+        if (tsLogger.logger.isDebugEnabled()) {
+            tsLogger.logger.debugf("%s - created, with expiry time of %s", getClass().getSimpleName(), EXPIRY_TIME);
+        }
+        this.recoveryStore = StoreManager.getRecoveryStore();
+        this.typeClass = (Class<T>) typeInstance.getClass();
+        this.typeName = typeInstance.type();
+        if (movedSubFolder != null && !movedSubFolder.isEmpty()) {
+            this.movedTypeName = this.typeName + movedSubFolder;
+        } else {
+            this.movedTypeName = null;
+        }
+    }
+
+    private boolean isNotNullUid(Uid uid) {
+        return !uid.equals(Uid.nullUid());
+    }
+
+    @Override
+    public void scan() {
+        Instant oldestSurviving = Instant.now().minus(EXPIRY_TIME);
+        if (tsLogger.logger.isDebugEnabled()) {
+            ZonedDateTime zdt = oldestSurviving.atZone(ZoneId.systemDefault());
+            tsLogger.logger.debugf("%s - scanning to remove items from before %s", getClass().getSimpleName(), TIME_FORMAT.format(zdt));
+        }
+        try {
+            InputObjectState uids = new InputObjectState();
+            if (this.recoveryStore.allObjUids(this.typeName, uids)) {
+                for (Uid uid = UidHelper.unpackFrom(uids); isNotNullUid(uid); uid = UidHelper.unpackFrom(uids)) {
+                    T action = this.typeClass.getConstructor(Uid.class).newInstance(uid);
+                    if (action.activate()) {
+                        Instant creationTime = Instant.ofEpochMilli(action.getCreationTimeMillis());
+                        if (creationTime.isBefore(oldestSurviving)) {
+                            try {
+                                if (this.movedTypeName != null) {
+                                    moveEntry(uid);
+                                } else {
+                                    removeEntry(uid);
+                                }
+                            } catch (Exception ex) {
+                                tsLogger.logger.warnf(ex, "%s - exception during attempted move of %s", getClass().getSimpleName(), uid);
+                            }
+                        }
+                    }
+                }
+            }
+        } catch (Exception ex) {
+            // end of uids!
```

Comment:
We'll end up here if the reflective access to the constructor fails so dumping the stack trace and reporting the error on System.out or the logger is necessary.

Reply:
It's the same thr original scanner works on error, maybe catch those Exceptions as another catch clause

Reply:
Replaced with multicatch and log message